lib/mayaUsd/ufe/UsdUndoMaterialCommands.cpp

```diff
@@ -375,43 +418,162 @@ void UsdUndoAssignNewMaterialCommand::redo()
         auto addMaterialCmd
             = std::dynamic_pointer_cast<MAYAUSD_NS::ufe::UsdUndoAddNewPrimCommand>(*cmdsIt++);
         auto addShaderCmd = std::dynamic_pointer_cast<UsdUndoCreateFromNodeDefCommand>(*cmdsIt);
-        connectShaderToMaterial(addShaderCmd->insertedChild(), addMaterialCmd->newPrim());
+        connectShaderToMaterial(addShaderCmd->insertedChild(), addMaterialCmd->newPrim(), _nodeId);
     }
 }
 
-void UsdUndoAssignNewMaterialCommand::connectShaderToMaterial(
-    Ufe::SceneItem::Ptr shaderItem,
-    UsdPrim             materialPrim)
+void UsdUndoAssignNewMaterialCommand::markAsFailed()
 {
```

Comment:
Ideas for future UFE: have a "Transactional" command compositor helper class that automtically undo + reset if not commited, to avoid having to sprinkle the "markAsFailed" everywhere. The reason is that if something throws an exception, markAsFailed won't be called and as far as I can tell, the already-executed commands won't be undone.

It is something I've been mulling over recently due to the bad interaction between composite command and create + execute.

Reply:
Yes, same issue in the current implementation of `Ufe::UndoableCommandGuard` which requires a setSuccess call to signify everything completed correctly, and fails to undo subtasks that were completed prior to the failure point. Designing a complete solution is hard and takes a significant time investment. Also, the use of exceptions adds a level of complexity since some clients like to swallow those and not propagate them upwards.